### PR TITLE
fix #3090 #3091 refining the store interface and hasSynced

### LINKED ISSUE #3013: IndexInformer cant restart after call function stop
@Override
  public void run() {
    if (started) {
      return;
    }

    started = true;

    this.processor.run();
    controllerThread.start();
  }

  @Override
  public void stop() {
    if (!started) {
      return;
    }

    stopped = true;
    controller.stop();
    controllerThread.interrupt();

    processor.stop();
  }

After call function stop；var "started" is still true;  so when call run() again, this just return after check "started" is true.

is this correct? or is a bug?

### LINKED ISSUE #3090: Informer.hasSynced and Store.isPopulated
A change in #1961 was to set hasSynced to false when re-listing.  This differs from the go client behavior - where once true, it won't flip back to false see https://github.com/kubernetes/client-go/blob/f6ce18ae578c8cca64d14ab9687824d9e1305a67/tools/cache/delta_fifo.go and the doc https://github.com/kubernetes/client-go/blob/f6ce18ae578c8cca64d14ab9687824d9e1305a67/tools/cache/shared_informer.go#L160

Can we return hasSynced to the original meaning and introduce more specific methods for additional checks - such as isWatching?

A change along these lines would allow us to eliminate the Store.isPopulated method

cc @rohanKanojia

### PR DESCRIPTION
**Pull request by shawkins**
## Description
Refining the Store interface by splitting internal and mutative methods to SyncableStore

Removing isPopulated and directly wiring the cache to be used
this adds an isRunning method to the SharedInformer.  this clarifies the
cache methods to be more like a map and the underlying map is now
concurrent to remove read locks

This could be seen as a breaking change since Store is somewhat of a public interface - however it seems much clearer and safer to have a narrower interface.

Closes #3013

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [x] Breaking change (fix or feature that would cause existing functionality to change
 - [x] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

**Comment by manusa** [comment]
Relates to: #2010

**Review by rohanKanojia — APPROVED** [review]

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/cache/Cache.java [hunk 1]
```diff
@@ -46,15 +47,15 @@
   private Map<String, Function<T, List<String>>> indexers = new HashMap<>();
 
   // items stores object instances
-  private Map<String, T> items = new HashMap<>();
+  private volatile ConcurrentHashMap<String, T> items = new ConcurrentHashMap<>();
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/cache/Cache.java:50]: Is it really needed for the ConcurrentHashMap to be declared as volatile?

Comment (manusa) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/cache/Cache.java:50]: Oh, I see it's being replaced later on

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Pull request merged**